File: 931-minimum-falling-path-sum/931-minimum-falling-path-sum.py
class Solution:
    # for loop instead and dynamic programming
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        m = len(matrix)
        n = len(matrix[0])
        if m == 1:
            return min(matrix[0])
        dp = [[None for _ in range(n)] for _ in range(2)]  # we only need two rows
        for j in range(n):
            dp[1][j] = matrix[m-1][j]
        for i in range(m-2, -1, -1):
            for j in range(n):
                min_val = sys.maxsize
                for k in range(max(0, j-1), min(n, j+2)):
                    min_val = min(min_val, dp[1][k])
                dp[0][j] = min_val + matrix[i][j]
            for j in range(n):
                dp[1][j] = dp[0][j]
        return min(dp[1])

    # RecursionError: maximum recursion depth exceeded in comparison
    """
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        m = len(matrix)
        n = len(matrix[0])
        dp = [[None for _ in range(n)] for _ in range(m)]
        for j in range(n):
            dp[m-1][j] = matrix[m-1][j]
        def recSearch(i: int, j: int) -> int:
            nonlocal dp
            if dp[i][j]:
                return dp[i][j]
            min_val = sys.maxsize
            for k in range(max(0, j-1), min(n, j+2)):
                min_val = min(min_val, recSearch(i, k))
            dp[i][j] = min_val + matrix[i][j]
            return dp[i][j]
        min_path = sys.maxsize
        for j in range(n):
            min_path = min(min_path, recSearch(0, j))
        return min_path
    """

File: 931-minimum-falling-path-sum/931-minimum-falling-path-sum.rs
use std::cmp;

impl Solution {
    pub fn min_falling_path_sum(matrix: Vec<Vec<i32>>) -> i32 {
        let m: i32 = matrix.len() as i32;
        let n: i32 = matrix[0].len() as i32;
        if (m == 1) {
            return *matrix[0].iter().min().unwrap();
        }
        let mut dp0: Vec<i32> = Vec::new();
        let mut dp1: Vec<i32> = Vec::new();
        for j in 0..n as usize {
            dp0.push(0);
            dp1.push(matrix[(m-1) as usize][j]);
        }
        let mut min_val: i32 = i32::MAX;
        for i in (0..(m-1)).rev() {
            for j in 0..n as usize {
                min_val = i32::MAX;
                for k in (cmp::max(0, (j-1) as i32))..(cmp::min(n, (j+2) as i32)) {
                    min_val = cmp::min(min_val, dp1[k as usize]);
                }
                dp0[j] = min_val + matrix[i as usize][j];
            }
            for j in 0..n as usize {
                dp1[j] = dp0[j];
            }
        }
        return *dp1.iter().min().unwrap();
    }
}
